### ISSUE TITLE
Rule L011 does not check capitalisation_policy config, always inserts 'AS'

### ISSUE THREAD

**Issue by NiallRees** [sqlfluff/sqlfluff#613]
```
select
    a.dog
from dogs a
```

always gets fixed to

```
select
    a.dog
from dogs AS a
```

It should check the capitalisation_policy to decide whether to insert an `AS` or `as`.

**Comment by barrywhart** [comment]
I ran `sqlfluff fix` on the above file. I got 3 linting violations, and the third one caused the alias to be removed.
```
L:   3 | P:  11 | L009 | Files must end with a trailing newline.
L:   3 | P:  11 | L011 | Implicit aliasing of table not allowed. Use explicit
                       | `AS` clause.
L:   3 | P:  14 | L031 | Avoid using aliases in join condition
```

This seems like a bug, since the query does not actually use `JOIN`. Perhaps this should be a new issue. Just noting it here for now.

**Comment by barrywhart** [comment]
I did some further investigation of this behavior...

As an experiment, I tried changing L011 to insert `"as"` rather than `"AS"`. This had no impact whatsoever on the output (it was capitalized regardless).

I believe what's happening is that L010 "wins", i.e. it sees the output of L01 and applies its own `capitalisation_policy` setting. In other words, the resulting `AS` will be upper or lower case depending on whether L010 `capitalisation_policy` is set to `upper` or `lower`.

Should this be considered a true bug? Or could it be considered a documentation issue?

This raises some potentially interesting questions on how various rules interact. I see that a couple of other rules already have a `capitalisation_policy`:
* `L014`
* `L030`

I wonder if any of the others are subject to conflicts like this.

**Comment by barrywhart** [comment]
@NiallRees and @alanmcruickshank: This issue is another example where multiple rules work together/depend on one another.

**Comment by jpy-git** [comment]
@NiallRees @barrywhart just had a look at this and yes L011 on it's own does not consider capitalisation. `AS` is a keyword so makes sense that it should be controlled by the capitalisation policy of L010. If L010 is also on then it works fine since that is the rule that cares about keyword capitalisation.

Obviously we could link L011 to the capitalisation policy of L010 so that it works on it's own. But my question there would be that the user is not including the rule (L010) which manages keyword capitalisation, so they can't really complain if there's inconsistent keyword capitalisation. If we did it for L011 we'd have to implement capitalisation check logic for any rule that potentially adds a keyword.

Basically don't think there's any interaction between the rules but rather that each rule should handle one thing and one thing only.
(and that L010 is one of those core formatting rules that doesn't ever really make sense to not have on)

**Comment by jpy-git** [comment]
Personally I'm of the opinion that L010 should be solely responsible for the capitalisation, but in the interest of providing a balanced argument:

I think if we insisted on having sensible AS capitalisation in L011 we can just say "find the first keyword in the parent stack and match it's capitalisation, and if it's not upper or lower then just use upper as default".

So we'd get things like:
```sql
select
    a.dog
from dogs as a
```
```sql
SELECT
    a.dog
FROM dogs AS a
```
```sql
seleCT
    a.dog
From dogs AS a
```
which makes sensible choices without overstepping it's bounds and trying to reimplement L010

**Comment by barrywhart** [comment]
I agree that rules shouldn't try and "embed" the behavior of other rules. I've made a similar argument for formatting (e.g. rule which splits long lines probably shouldn't be responsible for indenting the newly formed line after splitting). This is one reason I'm glad we're making progress on defining a set of "core rules", ones which we expect (although don't _require_ that are always enabled).

When I said there was interaction between the rules, what I really meant was, the user was complaining about rule X, when they should really be using rules X and Y. The rules only interact in the sense that they act on the same segments, and the combined effect of both rules is the desired behavior, one rule on its own may not produce a "nice" result.

**Comment by jpy-git** [comment]
Nice, glad we're on the same page 😄
